### PR TITLE
Add exploit for CVE-2019-1621, Cisco Data Center Network Manager arbitrary file download.

### PR DESCRIPTION
**Pull request by pedrib**
DCNM exposes a servlet to download files on /fm/downloadServlet.
      An authenticated user can abuse this servlet to download arbitrary files as root by specifying
      the full path of the file.
      This module was tested on the DCNM Linux virtual appliance 10.4(2), 11.0(1) and 11.1(1), and should
      work on a few versions below 10.4(2). Only version 11.0(1) requires authentication to exploit
      (see References to understand why).

See also:
https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20190626-dcnm-bypass
https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20190626-dcnm-file-dwnld

I will also put a full disclosure post up soon that contains more details, and will add the link here.

### CODE REVIEW

**Comment by wvu-r7** [comment]
Thanks, @pedrib!

**Comment by pedrib** [comment]
@wvu-r7 you're welcome, full disclosure link added!

**Comment by pedrib** [comment]
Corrected the title, this is actually the auxiliary arb download module; the file upload leading to RCE is PR  #12058

File: modules/auxiliary/admin/cisco/cisco_dcnm_download.rb [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+  def run
+    res = send_request_cgi({
+      'uri'    => normalize_uri(datastore['TARGETURI'], 'fm', 'fmrest', 'about','version'),
+      'method' => 'GET'
+    })
+    noauth = false
+
+    if res && res.code == 200
+      if res.body =~ /\"version\":\"11\.1\(1\)/
```
Comment (acammack-r7) [inline modules/auxiliary/admin/cisco/cisco_dcnm_download.rb]: See https://github.com/rapid7/metasploit-framework/pull/12058/files#r302228456 about using regex here.

**Comment by pedrib** [comment]
@acammack-r7 @wvu-r7 give me some time to address your points, I'm currently busy with other work! I don't like to make changes without testing locally with all the versions I have here (10.4.2, 11.0.1, 11.1.1), so need to downtime to address them. Please be patient!

**Comment by pedrib** [comment]
@acammack-r7 @wvu-r7 all done, good to go!

**Comment by pedrib** [comment]
ping!

**Comment by pedrib** [comment]
yo guys!

**Comment by wvu-r7** [comment]
And this one!

**Comment by pedrib** [comment]
@wvu-r7 please do, these are ready to go!

**Comment by pedrib** [comment]
The build failure is unrelated to the module I believe.

**Comment by pedrib** [comment]
Do you want a pcap to accelerate this?

**Comment by wvu-r7** [comment]
I might have to transfer these PRs to someone else. This might be my busiest week of the year. :(

**Comment by wvu-r7** [comment]
A pcap would help, yes, but is there a virtual appliance we can test? This could benefit from [module documentation](https://github.com/rapid7/metasploit-framework/wiki/Writing-Module-Documentation).

**Comment by wvu-r7** [comment]
Note to committers: please handle #12058 as well.

**Comment by pedrib** [comment]
> A pcap would help, yes, but is there a virtual appliance we can test? This could benefit from [module documentation](https://github.com/rapid7/metasploit-framework/wiki/Writing-Module-Documentation).

There is a virtual appliance, but you need a Cisco contract to access it. I'll send a pcap soon.

**Comment by wvu-r7** [comment]
Let me review this right quick, and if the pcap checks out, let's ship it.

**Review by wvu — CHANGES_REQUESTED** [review]
Quick-pass review. Should be clean enough to merge after.

**Comment by pedrib** [comment]
[dcnm_download.pcap.zip](https://github.com/rapid7/metasploit-framework/files/3553753/dcnm_download.pcap.zip)

Pcap attached

File: modules/auxiliary/admin/cisco/cisco_dcnm_download.rb [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+
+  def run
+    res = send_request_cgi({
+      'uri'    => normalize_uri(datastore['TARGETURI'], 'fm', 'fmrest', 'about','version'),
+      'method' => 'GET'
+    })
+    noauth = false
+
+    if res && res.code == 200
```
[the remark this answers is not in the document]
Comment (pedrib) [reply modules/auxiliary/admin/cisco/cisco_dcnm_download.rb]: Honestly I prefer to leave it as is, guard clauses make the code unreadable IMO.

File: modules/auxiliary/admin/cisco/cisco_dcnm_download.rb [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      if res.body.include?('version":"11.1(1)')
```
[the remark this answers is not in the document]
Comment (pedrib) [reply modules/auxiliary/admin/cisco/cisco_dcnm_download.rb]: Perhaps, but it's just 3 cases, and it's well tested this way, I prefer to keep it as is.

**Comment by pedrib** [comment]
Fixed most of your changes, except the ones where I provide a justification

**Comment by wvu-r7** [comment]
Cool, I can accept your justifications. Thanks for the quick changes!

**Comment by pedrib** [comment]
Done!

**Comment by wvu-r7** [comment]
Once https://github.com/rapid7/metasploit-framework/pull/12058#issuecomment-526033371 is addressed, I plan to review the pcaps and land both PRs.

**Comment by pedrib** [comment]
ok that's it!

**Comment by wvu-r7** [comment]
@pedrib: This one has an f-d reference. Okay to land?

**Comment by pedrib** [comment]
@wvu-r7 all good for all modules!

**Review by wvu — APPROVED** [review]

**Comment by wvu-r7** [comment]
# Release Notes

This adds a Cisco Data Center Network Manager arbitrary file download auxiliary module.

**Pull request merged**